### PR TITLE
Add dispatch based on compute architecture

### PR DESCRIPTION
**Pull request by ahendriksen**
This PR improves the ability to do dispatch based on compute architecture. It is a follow up to #1142.

It has two goals: 

1. Make it easier to specify which compute architectures a kernel is compatible with / should be compiled for.
2. Make it easier to compile a kernel only for the architectures for which it is used (if it is unused, the kernel should be empty).

We have a specific use case in RAFT for this feature. For the L2 pairwise distance kernel we have a CUTLASS based implementation that works om SM80+ and a fallback kernel. Preferably, each kernel is only compiled for the architectures on which it is actually used.

### CODE REVIEW

**Comment by ahendriksen** [comment]
This is a retry of PR #1295, which got inadvertently closed. I have taken into account the reviews that were performed there. In addition, I have centralized the dispatching more than in the previous PR. I am confident that this structure will support adding Hopper kernels without compiling duplicate and redundant kernels.

**Review by tfeher — APPROVED** [review]
Thanks Allard for the PR! I find the new helpers for dispatching based on compute arch useful, and overall the PR looks good to me.

It might make sense to shorten the PR description by putting description starting from "The previous approach can be described ..." in a separate comment.

**Comment by ahendriksen** [comment]
Thanks @tfeher! Moved PR description below:

The previous approach can be described as follows (using psuedo code):
``` c++
template <typename DistanceOpT>
__global__ void generic_kernel(DistanceOpT op, other arguments .. ) { .. implementation .. }

void cutlass_launch(args.. ) {.. implementation }

template <typename DistanceOpT>
__global__ void generic_kernel_prior_to_sm80(DistanceOpT op, args..) {
#if __CUDA_ARCH__ < 800
  .. copy and paste generic_kernel implementation
#endif
}

template <typename DistanceOpT>
void dispatch(DistanceOpT op, args..) {
  if (! op == l2_distance_op) {
    // run normal generic kernel
    generic_kernel(op, args..);
  } else {
    if (device_capability() >= 800) {
      cutlass_launch(args...);
    } else {
      generic_kernel_prior_to_sm80<<<grid, block, ...>>>(op, args...);
    }
  }
}
```
The main issue is that the generic kernel is copy pasted. In the new approach, the compute architectures for which the generic kernel has to be compiled (the "compatibility range") is given as an argument (using a compile-time tag type). This allows comparing to the compatibility range inside the kernel and early exiting if the architecture for which it is currently compiling is not supported. Therefore, we do not need to copy and paste generic kernel.

``` c++
template <typename DistanceOpT, typename SM_compat_t>
__global__ void generic_kernel(DistanceOpT op, SM_compat_t sm_compat_range, args .. ) {
  // Early exit to minimize the size of the kernel when it is not supposed to be compiled.
  if constexpr(! sm_compat_range.contains(raft::arch::SM_compute_arch())) {
    assert(false);
    return;
  }
  .. rest of implementation ..
}

void cutlass_launch(args.. ) {.. implementation }

template <typename DistanceOpT>
void dispatch(DistanceOpT op, args..) {
  if (! op == l2_distance_op) {
    // run normal generic kernel and compile for all architectures:
    auto full_range = raft::arch::SM_range(raft::arch::SM_min(), raft::arch::SM_future());
    generic_kernel<<<grid, block, ...>>>(op, full_range, args..);
  } else {
    // Get current architecture of device at runtime
    void* kernel_ptr = generic_kernel<OpT, decltype(legacy_range), ...>;
    auto runtime_arch = raft::arch::kernel_runtime_arch(kernel_ptr);
    // Define compatibility ranges for the cutlass and generic kernel
    auto cutlass_range = raft::arch::SM_range(raft::arch::SM_80(), raft::arch::SM_future());
    auto legacy_range = raft::arch::SM_range(raft::arch::SM_min(), raft::arch::SM_80());

    if (cutlass_range.contains(runtime_arch)) {
      // On SM80+: run cutlass kernel. (this might actually also compile a non-trivial kernel
      // for SM70 but that is unfortunately outside our control)
      cutlass_launch(_args...);
    } else {
      // This will run on architectures < SM80. Also, the compiled kernels from
      // SM80 and higher will be empty (< 10 instructions).
      generic_kernel<<<grid, block, ....>>>(op, legacy_range, args...);
    }
  }
}
```

File: cpp/include/raft/distance/detail/pairwise_matrix/dispatch.cuh [hunk 1]
```diff
@@ -104,114 +45,46 @@ void distance_matrix_dispatch(OpT distance_op,
...
+    auto runtime_arch = raft::arch::kernel_runtime_arch(kernel_ptr);
+
+    if (cutlass_range.contains(runtime_arch)) {
+      // If device is SM_80 or later, use CUTLASS-based kernel.
+      pairwise_matrix_sm80_dispatch(distance_op, params, cutlass_range, stream);
+    } else {
+      // Reuse kernel wrapper that we obtained above. This avoids performing the
+      // dispatch twice.
+      sm60_wrapper.launch(distance_op, params, stream);
+    }
```
[the remark this answers is not in the document]
Comment (ahendriksen) [reply cpp/include/raft/distance/detail/pairwise_matrix/dispatch.cuh:86]: I agree that there are a lot of different concepts and that this is the gnarliest part of the PR. The nice thing is that we concentrated all the gnarliest logic in one place. Before, it was in many different places. Also, I hope that we can soon drop the check against CTK V12. That will also make this code a lot more pleasant.

File: cpp/include/raft/util/arch.cuh [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+namespace raft::arch {
```
Comment (cjnolet) [inline cpp/include/raft/util/arch.cuh:18]: For new utility files, I'd like to start using `raft::util` namespace prefix to make them consistent w/ the other namespaces. At some point, we're going to scrape through the other utility files as well. `raft::util::arch` would be great here.

**Review by cjnolet — APPROVED** [review]
I'd like the namespace in `raft/util/arch.cuh` to match the nesting `raft::util::arch` so that we can start making util consistent w/ the other APIs in RAFT. I'm okay doing that in https://github.com/rapidsai/raft/pull/1307 though so that we can get this one merged.

**Comment by cjnolet** [comment]
@ahendriksen it  looks like there's a few CI failures for distance tests

**Comment by ahendriksen** [comment]
Hi Corey, yes there was a problem with column major input on Ampere. It took a bit longer to fix than expected. But I expect that the last commit will fix the issue :)

**Pull request merged**